crates/test-macro/src/lib.rs

```diff
@@ -34,23 +35,95 @@ pub fn wasm_bindgen_test(
     let mut leading_tokens = Vec::new();
     while let Some(token) = body.next() {
         leading_tokens.push(token.clone());
-        if let TokenTree::Ident(token) = token {
-            if token == "async" {
-                r#async = true;
+
+        match token {
+            // Parse a potential `should_panic` attribute
+            // Start by parsing the `#`
+            TokenTree::Punct(op) if op.as_char() == '#' => {
+                // We need to peek into the next token, save it before we continue.
+                let token = if let Some(token) = body.next() {
+                    leading_tokens.push(token.clone());
+                    token
+                } else {
+                    continue;
+                };
+
+                // Parse `[...]`
+                let group = match token {
+                    TokenTree::Group(group) if group.delimiter() == Delimiter::Bracket => group,
+                    _ => continue,
+                };
+
+                let mut stream = group.stream().into_iter();
+
+                // Parse `should_panic`
+                match stream.next() {
+                    Some(TokenTree::Ident(token)) if token == "should_panic" => (),
+                    _ => continue,
+                }
+
+                should_panic = Some(None);
+
+                // We are interested in the `expected` attribute or string if there is any
+                match stream.next() {
+                    // Parse the `(...)` in `#[should_panic(...)]`
+                    Some(TokenTree::Group(group))
+                        if group.delimiter() == Delimiter::Parenthesis =>
+                    {
+                        let mut stream = group.stream().into_iter();
```

Comment:
Ah, of course!

I would like to add failure tests with `trybuild`, what do you think?

Reply:
> I would like to add failure tests with `trybuild`, what do you think?

Yep, that's a good idea. They should probably go in the `wasm-bindgen-test-macro` crate, in line with how the existing ui tests for `#[wasm_bindgen]` are in the `wasm-bindgen-macro` crate.

Reply:
I actually found that I wasn't able to write a test with `should_panic` that fails, because apparently Rust parses `should_panic` in any case and will throw an error if you write it wrongly.

I will still include the tests in case Rust changes it's behavior.

Reply:
Done.

Reply:
Apparently `wasm-bindgen-test` doesn't support 1.56, so I bumped CI to 1.59.
I hope that's alright, see https://github.com/rustwasm/wasm-bindgen/issues/3114.

Reply:
Interestingly enough I can't reproduce the exact output of trybuild locally.
In the end I just copied the output from CI.